### PR TITLE
ci: integrate with Codecov for test coverage

### PR DESCRIPTION
**Pull request by Jusshersmith**
## Problem

We already try hard to make sure code has appropriate tests written, however we don't currently provide an easy way to monitor and track test coverage.

## Solution

Integrate the sso repo with a tool like [Codecov](https://codecov.io/), which helps us track and monitor coverage reports and make them a prominent part of pull requests and the contribution workflow.

## Notes

This currently uses the Codecov [bash uploader](https://docs.codecov.io/v4.3.0/docs/about-the-codecov-bash-uploader) to upload coverage reports, which is one of the recommended methods. Other methods include updating CircleCI to version 2.1 and making use of CircleCI [orbs](https://circleci.com/blog/making-code-coverage-easy-to-see-with-the-codecov-orb/).

Specific settings for codecov to be decided - this PR has been created with defaults.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -27,9 +27,13 @@ jobs:
...
+          name: push sso-dev image
+          command: |
+            if [[ -n $DOCKER_USER ]]; then
+              docker login -u $DOCKER_USER -p $DOCKER_PASS
+              make imagepush
+            fi
+      - run:
+          name: upload coverage file to codecov
+          command: |
+            bash <(curl -s https://codecov.io/bash)
```
Comment (Jusshersmith) [inline .circleci/config.yml]: If we stick to this method, would we prefer to cache this file in CircleCI, or host it locally?

File: .codecov/codecov.yml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      conditional: yes
+      loop: yes
+      method: no
+      macro: no
+
+comment:
+  layout: "header, diff"
+  behavior: default
+  require_changes: no
+
```
Comment (Jusshersmith) [inline .codecov/codecov.yml]: This is largely default, with the only changes so far being:
- removal of notifications for the time being
- lowered the 'range' (preferred range, among other settings to be decided)

File: .circleci/config.yml (hunk 1 above)
Comment (Jusshersmith) [inline .circleci/config.yml]: If we stick to this method, would we prefer to cache this file in CircleCI, or host it locally?

File: .codecov/codecov.yml (hunk 2 above)
Comment (jphines) [reply .codecov/codecov.yml]: :+1:

File: .circleci/config.yml (hunk 1 above)
Comment (jphines) [reply .circleci/config.yml]: This seems like the suggested method, right? let's just do this for now.  The next logical step would be to orb it.

**Comment by jphines** [comment]
What do we need to merge this @Jusshersmith ?

**Comment by Jusshersmith** [comment]
@jphines I think it's fine to merge as is - we'll work out the exact configuration we prefer as time goes on, but everything seems to be working well enough to start with! Also once this is merged I think the reports become more useful as they'll have master to compare against.

File: .codecov/codecov.yml (hunk 2 above)
Comment (Jusshersmith) [reply .codecov/codecov.yml]: Also added a list of files that are impacted by the pull request and their coverage to the PR comment.

**Pull request merged**